File: sw_python/simulate.py
# -*- coding: utf-8 -*-
import numpy as np

def simulate_gr4j(nStep, x1, x2,x3,x4, upperTankRatio, lowerTankRatio, maxDayDelay, UH1, UH2, Pn, En):
    # 定义一些产汇流计算需要用到的数值及数组
    S0 = upperTankRatio * x1            # 定义产流水库初识土壤含水量 = 初识土壤含水比例 * 产流水库容量
    R0 = lowerTankRatio * x3            # 定义汇流水库初始土壤含水量 = 初识土壤含水比例 * 汇流水库容量
    S_TEMP = S0                         # 用S_temp存储当前产流水库储量
    R_TEMP = R0                         # 用R_TEMP存储当前汇流水库储量

    S = np.zeros(nStep)                 # 产流水库数组，每天的土壤含水量
    R = np.zeros(nStep)                 # R：汇流水库逐日水量
    Ps = np.zeros(nStep)                # 产流中间变量，指每天有多少水补充土壤含水量
    Es = np.zeros(nStep)                # 产流中间变量，指每天土壤含水量还有多少消耗需要用来蒸发
    Perc = np.zeros(nStep)              # 产流中间变量，指每天土壤含水量通过下渗产生多少直接径流
    Pr = np.zeros(nStep)                # GR4j 总产流量

    UH_Fast = np.zeros((nStep, maxDayDelay))  # UH_Fast: 用于记录UH1单位线作用下的产流信息
    UH_Slow = np.zeros((nStep, maxDayDelay * 2))  # UH_Slow: 用于记录UH2单位线作用下的产流信息
    F = np.zeros(nStep)
    Qr = np.zeros(nStep)  # Qr：汇流水库快速流出流量
    Qd = np.zeros(nStep)  # Qd：汇流水库慢速流出流量
    Q = np.zeros(nStep)  # Q：汇流总出流量

    for i in range(nStep):
        S[i] = S_TEMP
        R[i] = R_TEMP

        # 使用Pn[i] 和 En[i]来进行判断
        if Pn[i] != 0:  # 净雨量大于0，此时一部分净雨形成地面径流，一部分下渗
            Ps[i] = x1 * (1 - ((S[i] / x1) ** 2)) * np.tanh(Pn[i] / x1) / (1 + S[i] / x1 * np.tanh(Pn[i] / x1))
            Es[i] = 0
        if En[i] != 0:  # 净蒸发能力大于0，此时土壤中水分一部分用于消耗
            Ps[i] = 0
            Es[i] = (S[i] * (2 - (S[i] / x1)) * np.tanh(En[i] / x1)) / (1 + (1 - S[i] / x1) * np.tanh(En[i] / x1))

        # 更新上层水库蓄水量
        S_TEMP = S[i] - Es[i] + Ps[i]
        Ratio_TEMP = S_TEMP / x1

        # 计算产流水库渗漏
        Perc[i] = S_TEMP * (1 - (1 + (4.0 / 9.0 * (S_TEMP / x1)) ** 4) ** (-0.25))

        # 计算总产流量
        Pr[i] = Perc[i] + (Pn[i] - Ps[i])

        # 更新当前产流水库水量，作为次日产流水库水量
        S_TEMP = S_TEMP - Perc[i]

        # 至此产流过程全部结束，进入汇流阶段
        # 汇流计算
        # 计算地表水与地下水之间的水量交换
        F[i] = x2 * (R[i] / x3) ** 3.5

        # 计算地表水汇流，将产流量按照90%(快速)和10%(慢速)划分
        # 快速地表径流汇流使用单位线UH1；慢速地表径流汇流使用单位线UH2
        R_Fast = Pr[i] * 0.9
        R_Slow = Pr[i] * 0.1

        if i == 0:
            UH_Fast[i, :] = R_Fast * UH1  # 第1时段产流量在时间上的分配
            UH_Slow[i, :] = R_Slow * UH2  # 第1时段产流量在时间上的分配
        else:
            UH_Fast[i, :] = R_Fast * UH1  # 先计算当前时段产流量在时间上的分配
            for j in range(maxDayDelay - 1):
                UH_Fast[i, j] = UH_Fast[i, j] + UH_Fast[i - 1, j + 1]  # 第2时段总汇流=第2时段产流量当前汇流+第1时段产流量第2部分汇流

            UH_Slow[i, :] = R_Slow * UH2  # 先计算当前时段产流量在时间上的分配
            for j in range(2 * maxDayDelay - 1):
                UH_Slow[i, j] = UH_Slow[i, j] + UH_Slow[i - 1, j + 1]  # 第2时段总汇流=第2时段产流量当前汇流+第1时段产流量第2部分汇流

        # 此处应注意下标
        # 更新汇流水库水量变化
        R_TEMP = max(0, R_TEMP + UH_Fast[i, 1] + F[i])

        # 计算汇流水库快速流出流量
        Qr[i] = R_TEMP * (1 - (1 + (R_TEMP / x3) ** 4) ** (-0.25))

        # 再次更新汇流水库水量变化
        R_TEMP = R_TEMP - Qr[i]

        # 此处应注意下标
        # 计算汇流水库慢速流出流量
        Qd[i] = max(0, UH_Slow[i, 1] + F[i])

        # 计算汇流总出流量
        Q[i] = Qr[i] + Qd[i]

    return Q

File: gr4j/src/main/java/com/sw/simulate.java
package com.sw;

/**
 * @author Marchino
 * @date 2024/6/3 14:26
 * @description
 */

public class simulate {
    public double[] simulateGr4j(int nStep, double x1, double x2, double x3, double x4, double upperTankRatio, double lowerTankRatio, int maxDayDelay, double[] UH1, double[] UH2, double[] Pn, double[] En){
        double S0 = upperTankRatio * x1;
        double tempS = S0;
        double[] S= new double[nStep];
        double[] Ps = new double[nStep];
        double[] Es = new double[nStep];
        double[] Pr = new double[nStep];
        double RFast, RSlow;
        double[][] UHFast = new double[nStep][maxDayDelay];
        double[][] UHSlow = new double[nStep][maxDayDelay];
        double tempR = 0;
        double[] Qr = new double[nStep];
        double[] Qd = new double[nStep];
        double[] Q = new double[nStep];
        double[] F = new double[nStep];
        double[] R = new double[nStep];

        double[] perc = new double[nStep];
        for (int i = 0; i < nStep; i++) {
            S[i] = tempS;
            R[i] = tempR;
            if (Pn[i] != 0) {
                Pn[i] = x1 * (1 - (Math.pow(S[i] / x1, 2))) * Math.tanh(Pn[i] / x1) / (1 + S[i] / x1 * Math.tanh(Pn[i] / x1));
                Es[i] = 0;
            }
            if (En[i] != 0) {
                Ps[i] = 0;
                Es[i] = (S[i] * (2 - (S[i] / x1)) * Math.tanh(En[i] / x1)) / (1 + (1 - S[i] / x1) * Math.tanh(En[i] / x1));
            }

            tempS = tempS - Es[i] + Ps[i];
            perc[i] = tempS * Math.pow(1 - (1 + (4.0 / 9.0 * Math.pow(tempS / x1, 4))), -0.25);
            Pr[i] = perc[i] + (Pn[i] - Ps[i]);

            tempS = tempS - perc[i];

            F[i] = Math.pow(x2 * (R[i] / x3), 3.5);

            RFast = 0.9 * Pr[i];
            RSlow = 0.1 * Pr[i];

            if (i == 0) {
                for (int j = 0; j < maxDayDelay; j++) {
                    UHFast[0][j] = RFast * UH1[j];
                    UHSlow[0][j] = RSlow * UH2[j];
                }
            } else {
                for (int j = 0; j < maxDayDelay - 1; j++) {
                    UHFast[i][j] = RFast * UH1[j];
                    UHFast[i][j] += UHFast[i - 1][j + 1];
                }
                for (int j = 0; j < maxDayDelay * 2 - 1; j++) {
                    UHSlow[i][j] = RSlow * UH2[j];
                    UHSlow[i][j] += UHSlow[i - 1][j + 1];
                }
            }

//            tempR = Math.max(0, tempR + UHFast[i, 1] + F[i]);
//            Qr[i] =  tempR * (1 - Math.pow(1 + Math.pow(tempR / x3, 4), -0.25 ));
//            Qd[i] = Math.max(0, UHSlow[i, 1] + F[i]);
//            Q[i] = Qr[i] + Qd[i];

            tempR = Math.max(0, tempR + UHFast[i][1] + F[i]);
            Qr[i] = tempR * (1 - Math.pow(1 + Math.pow(tempR / x3, 4), -0.25));
            Qd[i] = Math.max(0, UHSlow[i][1] + F[i]);
            Q[i] = Qr[i] + Qd[i];


        }

        return Q;
    }
}
